ledger/src/blockstore_processor.rs

```diff
@@ -976,14 +976,22 @@ fn load_frozen_forks(
                     ).and_then(|supermajority_root| {
                         if supermajority_root > *root {
                             // If there's a cluster confirmed root greater than our last
-                            // replayed root, then beccause the cluster confirmed root should
+                            // replayed root, then because the cluster confirmed root should
                             // be descended from our last root, it must exist in `all_banks`
                             let cluster_root_bank = all_banks.get(&supermajority_root).unwrap();
 
                             // cluster root must be a descendant of our root, otherwise something
                             // is drastically wrong
                             assert!(cluster_root_bank.ancestors.contains_key(root));
                             info!("blockstore processor found new cluster confirmed root: {}, observed in bank: {}", cluster_root_bank.slot(), bank.slot());
+
+                            // Ensure cluster-confirmed root is set as root in blockstore
+                            if !blockstore.is_root(supermajority_root) {
```

Comment:
I think the fix should be to iterate from the last `max_root` (which is known to be rooted in blockstore) to the new `supermajority_root` and set root for all those. This can be done by iterating through the `bank.parent` starting from supermajority root until we find the previous `max_root`, then calling `set_root()` on all of those slots

Reply:
and also to guarantee that `max_root` starting from the very beginning is always a root in blockstore, we should also set root after this: https://github.com/solana-labs/solana/blob/master/ledger/src/blockstore_processor.rs#L910 in case `root_bank > blockstore_max_root`, i.e. we grabbed a network snapshot ahead of our last known root.

Reply:
I also think we should call `blockstore.set_root()` on everything between `blockstore_max_root` and `root_bank.slot()`  via iterating over `AncestorIterator::new(max_root, blockstore)` here: https://github.com/solana-labs/solana/blob/master/ledger/src/blockstore_processor.rs#L910 in case the snapshot `root_bank.slot() > blockstore_max_root` (we received a greater snapshot from the network). If there's a gap in the ledger here, i.e. `0->1->2 (blockstore_max_root) -> 3 -> 4 -> 5 (grabbed snapshot from network)`, then because `3` and `4` are missing we may never find `blockstore_max_root`, but maybe that's an acceptable failure condition

Reply:
but from these two guarantees above, inductively it should be true that just by setting the roots for the ancestors between `*root` and `supermajority_root` on every iteration, then we won't miss any roots in blockstore, and we don't need the `if !blockstore.is_root(supermajority_root)` check.

Reply:
> I also think we should call `blockstore.set_root()` on everything between `blockstore_max_root` and `root_bank.slot()`

Hmm, that might be ultimately useful, but I'd prefer not to do this at this time. Currently services like rpc expect slots skipped due to boot from newer snapshot to be completely absent in blockstore. I'd want to examine the ramifications of those slots appearing in the Roots column but not elsewhere.

Reply:
We concluded over DM to put off backfilling blockstore roots when `root_bank.slot() > blockstore_max_root` until we can examine behavior changes more closely.